app/services/report_importer.rb

```diff
@@ -2,14 +2,38 @@ class ReportImporter
   delegate :logger, :to => :Rails
   attr_reader :report
 
+  def self.importer_for(type)
+    importers[type.to_s] || importers[:puppet]
+  end
+
+  def self.importers
+    @importers ||= { :puppet => ConfigReportImporter }.with_indifferent_access
+  end
+
+  def self.register_report_importer(key, klass)
+    importers[key.to_sym] = klass
+  end
+
+  def self.fact_features
+    importers.map { |_type, importer| importer.authorized_smart_proxy_features }.compact.flatten.uniq
+  end
+
+  def self.authorized_smart_proxy_features
+    Rails.logger.debug("Importer #{self} does not implement authorized_smart_proxy_features.")
+    Foreman::Deprecation.deprecation_warning(1.12, "ReportImporter is a superclass, please use child classes. Maybe try: ConfigReportImporter?")
+    ['Puppet', 'Chef Proxy']
+  end
+
   def self.import(raw, proxy_id = nil)
     importer = new(raw, proxy_id)
     importer.import
     importer.report
   end
 
-  def self.report_features
-    ['Puppet', 'Chef Proxy']
+  # to be defined in children
+  def report_name_class
+    Foreman::Deprecation.deprecation_warning(1.12, "ReportImport is a supperclass, please use child classes to define 'report_class_name'")
```

Comment:
1.12 should be string I think, also I'm wondering if any plugin uses this class directly, maybe we don't need to keep backwards compatibility here